src/lib/statistics/cardinality_estimator.cpp

```diff
@@ -0,0 +1,858 @@
+  if (const auto value_expression = std::dynamic_pointer_cast<ValueExpression>(limit_node.num_rows_expression())) {
+    const auto num_rows = Cardinality{type_cast_variant<float>(value_expression->value)};
+    auto column_statistics = std::vector<std::shared_ptr<BaseColumnStatistics>>{limit_node.column_expressions().size()};
+
+    for (auto column_id = ColumnID{0}; column_id < input_table_statistics->column_statistics.size(); ++column_id) {
+      resolve_data_type(input_table_statistics->column_data_type(column_id), [&](const auto data_type_t) {
+        using ColumnDataType = typename decltype(data_type_t)::type;
+        column_statistics[column_id] = std::make_shared<ColumnStatistics<ColumnDataType>>();
+      });
+    }
+
+    return std::make_shared<TableStatistics>(std::move(column_statistics), num_rows);
+  } else {
+    return input_table_statistics;
+  }
+}
+
+std::shared_ptr<TableStatistics> CardinalityEstimator::estimate_operator_scan_predicate(
+    const std::shared_ptr<TableStatistics>& input_table_statistics, const OperatorScanPredicate& predicate) {
+  /**
+   * This function analyses the `predicate` and dispatches an appropriate selectivity-estimating algorithm.
+   */
+
+  auto selectivity = Selectivity{1};
+
+  const auto left_column_id = predicate.column_id;
+  auto right_column_id = std::optional<ColumnID>{};
+
+  const auto left_input_base_column_statistics = input_table_statistics->column_statistics[left_column_id];
+  const auto left_data_type = input_table_statistics->column_data_type(left_column_id);
+
+  auto output_column_statistics =
+      std::vector<std::shared_ptr<BaseColumnStatistics>>{input_table_statistics->column_statistics.size()};
+
+  resolve_data_type(left_data_type, [&](const auto data_type_t) {
+    using ColumnDataType = typename decltype(data_type_t)::type;
+
+    const auto left_input_column_statistics =
+        std::static_pointer_cast<ColumnStatistics<ColumnDataType>>(left_input_base_column_statistics);
+
+    /**
+     * Estimate IS (NOT) NULL
+     */
+    if (predicate.predicate_condition == PredicateCondition::IsNull) {
+      const auto null_value_ratio =
+          estimate_null_ratio_of_column(*input_table_statistics, *left_input_column_statistics);
+
+      if (null_value_ratio) {
+        selectivity = *null_value_ratio;
+
+        // All that remains of the column we scanned on are NULL values
+        const auto column_statistics = std::make_shared<ColumnStatistics<ColumnDataType>>();
+        column_statistics->null_value_ratio = std::make_shared<NullValueRatioStatistics>(1.0f);
+        output_column_statistics[left_column_id] = column_statistics;
+      } else {
+        // If have no null-value ratio available, assume a selectivity of 1, for both IS NULL and IS NOT NULL
+        selectivity = 1.0f;
+      }
+    } else if (predicate.predicate_condition == PredicateCondition::IsNotNull) {
```

Comment:
this looks like a lot of duplication for something that is essentially `1.0 - x`.

Reply:
merged those branches into one